### PR TITLE
ChoiceLayer: return length normalized score as expected

### PR DESCRIPTION
**Pull request by patrick-wilken**
Currently the search score depends on the batch and beam size when doing length normalization. This is because the implementation of length normalization in ChoiceLayer introduces a factor of maximum decoding length for all scores, which can vary depending on which sequences are in the batch/beam.

This PR removes the factor, such that ChoiceLayer returns the expected score/seq_lens once search is done.

### CODE REVIEW

File: TFNetworkRecLayer.py [hunk 1]
```diff
@@ -2600,7 +2600,9 @@ def search_resolve_cond(i, *args):
       acc_search_choices = SearchChoices(owner=latest_layer_choice, beam_size=latest_beam_size)
       final_choice_rec_vars = self.get_layer_rec_var_from_loop_vars(
         loop_vars=final_net_vars,
-        layer_name=latest_layer_choice.name)
+        layer_name=latest_layer_choice.name,
+        final_frame=True,
```
Comment (patrick-wilken) [inline TFNetworkRecLayer.py]: This should have been set before, right? Or am I missing something?

File: TFNetworkRecLayer.py [hunk 2]
```diff
@@ -4770,6 +4773,20 @@ def get_dep_layers(self):
...
+    """
+    :param dict[str,tf.Tensor] rec_vars_outputs:
+    :param tf.Tensor seq_len: shape (batch,)
+    :rtype: dict[str,tf.Tensor]
+    """
+    if self.length_normalization:
+      assert "choice_scores" in rec_vars_outputs
+      # Finalize length normalization. During search we keep an extra factor t (recurrent time step) for efficiency
+      # reasons (see self.get_output()). Remove it here.
+      num_time_steps = tf.reduce_max(seq_len) + 1  # + 1 to include sequence end
```
Comment (albertz) [inline TFNetworkRecLayer.py:4786]: Note that `+1` is not correct in all cases. You can configure whether the seq/RecLayer includes EOS or not.

File: TFNetworkRecLayer.py (hunk 1 above)
Comment (albertz) [reply TFNetworkRecLayer.py:2604]: Hm yes, good question. Not sure anymore. But looks right. I'm not 100% sure about `seq_len`, if that comes from another beam.

**Comment by albertz** [comment]
I will just merge it now. It's already an improvement. We can improve upon it more later.
